projects/common/src/telemetry/user-telemetry-internal.service.ts

```diff
@@ -0,0 +1,144 @@
+import { Injectable, Injector } from '@angular/core';
+import { NavigationEnd, Router } from '@angular/router';
+import { Observable, ReplaySubject, Subject } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { Dictionary } from '../utilities/types/types';
+import { UserTelemetryProvider, UserTelemetryRegistrationConfig, UserTraits } from './telemetry';
+
+@Injectable({ providedIn: 'root' })
+export class UserTelemetryInternalService {
+  private telemetryProviders: UserTelemetryInternalConfig[] = [];
+  private readonly telemetryActionSubject: Subject<TelemetryAction> = new ReplaySubject();
+  private readonly telemetryAction$: Observable<TelemetryAction>;
+  private readonly identifyAction$: Observable<TelemetryAction>;
+  private readonly trackEventAction$: Observable<TelemetryAction>;
+  private readonly trackPageAction$: Observable<TelemetryAction>;
+  private readonly trackErrorAction$: Observable<TelemetryAction>;
+
+  public constructor(private readonly injector: Injector, private readonly router: Router) {
+    this.telemetryAction$ = this.telemetryActionSubject.asObservable();
+
+    this.identifyAction$ = this.telemetryAction$.pipe(filter(action => action.type === TelemetryActionType.Identify));
+
+    this.trackEventAction$ = this.telemetryAction$.pipe(
+      filter(action => action.type === TelemetryActionType.TrackEvent)
```

Comment:
I still maintain that using observables here just adds complexity for no benefit, but it's internal so not a big deal.

Reply:
Is it complicating the code? I agree it probably makes a small perf improvement, but theoretically, the browser should spend as few cycles as possible in the original execution loop. Let's say we start tracking the `mouseover` event for topology (since we show some contents over there), then this additional processing may have a more perceived impact.

Reply:
I think so as a reader (and I'm more comfortable with observables than most!) We're basically just splitting one minuscule event into two minuscule events - the call to this function, which pushes into the observable and then returns control, and then the subscription to that observable which pushes into each telemetry provider. Inside the telemetry providers, they're almost certainly designed to be very performant - either dropping (in a separate event) an async call to report the event, or storing it in a buffer for a later report. Because the foreach is on subscriber side, all of those calls to each provider still happen as part of one event anyway - and since we're running off one event loop (if I remember my JS, which admittedly is super rusty), any delay there would still be user perceived, same as if we had not used the observable - it's not running in a different thread - but we're talking so far outside the realm of perceived perf, it really shouldn't be relevant. Happy to be proved wrong though!

Reply:
One thing which is concerning is that we are loading these telemetry platforms through Js during the later stage of app initialization (it is not loaded via index.html). During this stage, I don't the telemetry tool to completely miss the first-page load event because it is not initialized. With an observable-based approach and a Replay subject, I can ensure that those initial events are not missed.

Reply:
`it's not running in a different thread ` I can use the `observeOn(async)` that would help.

In general, I agree it is verbose. But lemme know what you think about the above concern.

Reply:
P.S I may have to make some edits to the existing observable code to take care of the above concern.

Reply:
> With an observable-based approach and a Replay subject, I can ensure that those initial events are not missed.

Yeah that's a fair point - although how do we know when the providers are ready to receive events?

> observeOn(async)

Not sure what that would help - rx uses async by default. My main point was that making things async doesn't just background things like it does in when you put it on a different thread. Anyway, it's moot if we want to support the delayed start case anyway.

Reply:
> Yeah, that's a fair point - although how do we know when the providers are ready to receive events?

Telemetry tools that come with an npm package, usually allow a callback in their init function, which gets called after the load and init is done. We could use that. 

This would need some changes to the current design. 

For tools with only js script (that we will host. For example. fresh paint), we can build a way to introduce a callback once the object is availabl.e

Reply:
So, probably we should have `initialize(config, callback: () => void)` ? And then in the orchestration service, we can use this before we start pushing the events to the tool

Reply:
I wouldn't work with callbacks when we're working with Rx everywhere else (especially since that was part of the argument for why it's Rx!). Perhaps pass in an observer?

Reply:
Lemme check

Reply:
So, I tried this but the whole design was getting complicated with too many subscriptions. I am reverting the whole observable approach for now. For a delayed start, I will first verify how these telemetry tools handle it. After that, I may bring back the observable approach if required.